Improve error message for incorrect store format

spacecowboy opened the pull request:
Be more explicit about how the user can open an enterprise-only store if
current config is set to community.

spacecowboy committed:
Improve error message for incorrect store format

Be more explicit about how the user can open an enterprise-only store if
current config is set to community.

lutovich commented on community/kernel/src/main/java/org/neo4j/kernel/impl/storemigration/UpgradableDatabase.java:
```diff
@@ -84,6 +87,13 @@ public RecordFormats checkUpgradeable( File storeDirectory )
 
+        // If we are trying to open an enterprise store when configured to use community format, then inform the user
+        // of the config setting to change since downgrades aren't possible but the store can still be opened.
+        if ( result.actualVersion.startsWith( "vE" ) && !format.storeVersion().startsWith( "vE" ) )
```
I think we should not check version strings like this. It is better to check upgradability somehow by looking at `fromFormat` and `format`. Maybe even add a method like `RecordFormats#canBeUpgradesFrom(RecordFormats)` and then create `Result` with outcome `unexpectedUpgradingStoreFormat`. Implementation of `#canBeUpgradesFrom()` can simply check string equality for now.

spacecowboy replied:
`result` IS `fromFormat`. Problem is that you can't instantiate the `fromFormat` object from result because the enterprise format is not a known format in this case, resulting in `IllegalArgumentException( "Unknown store version '" + storeVersion + "'" )` (see `RecordFormatSelector.selectForVersion( result.actualVersion )`).

I can move the if-clause above into `RecordFormatSelector` if you prefer, e.g. something like:

In `RecordFormatSelector`:

``` java
    /**
     * Select record formats for provided store version.
     *
     * @param storeVersion store version to find format for
     * @return record formats
     * @throws IllegalArgumentException if format for specified store version not found
     */
    public static RecordFormats selectForVersion( String storeVersion )
    {
        Iterable<RecordFormats> currentFormats =
                map( RecordFormats.Factory::newInstance, Service.load( RecordFormats.Factory.class ) );
        for ( RecordFormats format : concat( KNOWN_FORMATS, currentFormats ) )
        {
            if ( format.storeVersion().equals( storeVersion ) )
            {
                return format;
            }
        }
        // If we are trying to open an enterprise store but this is not a known format, 
        // e.g. we are running community, then inform the user
        // of the config setting to change since downgrades aren't possible but the store can still be opened.
        if ( storeVersion.startsWith( "vE" ) )
        {
            throw new StoreUpgrader.UnexpectedUpgradingStoreFormatException();
        }
        throw new IllegalArgumentException( "Unknown store version '" + storeVersion + "'" );
    }
```

What do you think @lutovich ?

lutovich replied:
Imo it is better to keep this check in `UpgradableDatabase` because `RecordFormatSelector` should just select format and not care about upgrade paths we support.

The main thing I do not like about this solution is `startsWith( "vE" )`. What do you think about introducing a class with store version constants in community like:

```
class StoreVersions {
    private static final String STANDARD_V2_0 = "v0.A.1";
    // ...
    private static final String STANDARD_V3_0 = "v0.A.7";
    private static final String HIGH_LIMIT = "vE.H.0";

    private StoreVersions() {}

    public static boolean isCommunityStoreVersion(String storeVersion) {
         // assert that storeVersion among those we know
         return STANDARD_V2_0.equals(storeVersion) || STANDARD_V3_0.equals(storeVersion);
    }

     public static boolean isEnterpriseStoreVersion(String storeVersion) {
         // assert that storeVersion among those we know
        return HIGH_LIMIT.equals(storeVersion);
     }
}
```

Then all `RecordFormats` implementors would reference these constants:

```
public class HighLimit extends BaseRecordFormats {
    // ...
    public static final String STORE_VERSION = StoreVersions.HIGH_LIMIT;
    // ...
}
```

and in `UpgradableDatabase` check would look like:

```
 if ( isEnterpriseStoreVersion(result.actualVersion) && isCommunityStoreVersion(format.storeVersion()) ) {...}
```

with this approach knowledge about store versions is centralized. Drawback, of course, is that we put constant related to enterprise-only format in community. We already do this though in description of `GraphDatabaseSettings#record_format`.

lutovich commented:
@spacecowboy sorry, forgot about this PR. made a comment now.

spacecowboy committed:
Add a list of all store format versions

Which allows for convenient querying if store format is community or
enterprise.
